File: songchord-ai/backend/app/services/sync.py
"""Synchronize each chord to the nearest lyric word (by start time)."""

from dataclasses import dataclass

from app.domain import Chord, LyricLine


@dataclass
class WordRef:
    line_index: int
    word_index: int
    start: float
    end: float


@dataclass
class ChordSync:
    chord: Chord
    line_index: int | None
    word_index: int | None


def _flatten_words(lines: list[LyricLine]) -> list[WordRef]:
    refs = []
    for li, line in enumerate(lines):
        for wi, word in enumerate(line.words):
            refs.append(WordRef(line_index=li, word_index=wi, start=word.start, end=word.end))
    return refs


def _nearest_word_ref(t: float, refs: list[WordRef]) -> WordRef | None:
    if not refs:
        return None
    containing = [r for r in refs if r.start <= t < r.end]
    if containing:
        return containing[0]
    return min(refs, key=lambda r: abs(r.start - t))


def sync_chords_to_words(chords: list[Chord], lines: list[LyricLine]) -> list[ChordSync]:
    refs = _flatten_words(lines)
    result = []
    for chord in chords:
        ref = _nearest_word_ref(chord.start, refs)
        result.append(
            ChordSync(
                chord=chord,
                line_index=ref.line_index if ref else None,
                word_index=ref.word_index if ref else None,
            )
        )
    return result

File: songchord-ai/frontend/src/lib/sync.ts
import type { Chord, LyricLine } from "./types";

/** Index of the lyric line that should be highlighted at the given playback time. */
export function activeLineIndex(lines: LyricLine[], currentTime: number): number {
  for (let i = 0; i < lines.length; i += 1) {
    if (currentTime >= lines[i].start && currentTime < lines[i].end) {
      return i;
    }
  }
  let lastStarted = -1;
  for (let i = 0; i < lines.length; i += 1) {
    if (lines[i].start <= currentTime) lastStarted = i;
  }
  return lastStarted;
}

export interface ChordPlacement {
  chord: Chord;
  chordIndex: number;
  lineIndex: number | null;
  /** 0..1 position within the line's [start, end) span, for proportional layout. */
  ratio: number | null;
}

/** Assign each chord to the lyric line active at its start time, mirroring the
 * backend's word-level sync but at line granularity (sufficient for on-screen
 * placement; the backend result already carries word-accurate timing). */
export function placeChordsOnLines(chords: Chord[], lines: LyricLine[]): ChordPlacement[] {
  return chords.map((chord, chordIndex) => {
    const lineIndex = activeLineIndex(lines, chord.start);
    if (lineIndex === -1) {
      return { chord, chordIndex, lineIndex: null, ratio: null };
    }
    const line = lines[lineIndex];
    const span = line.end - line.start;
    const ratio = span > 0 ? Math.min(1, Math.max(0, (chord.start - line.start) / span)) : 0;
    return { chord, chordIndex, lineIndex, ratio };
  });
}
